diff --git a/src/lib_protocol_compiler/compiler.ml b/src/lib_protocol_compiler/compiler.ml
index 729c8d1cd0..ad4a433012 100644
--- a/src/lib_protocol_compiler/compiler.ml
+++ b/src/lib_protocol_compiler/compiler.ml
@@ -38,9 +38,16 @@ let preloaded_cmis : Persistent_env.Persistent_signature.t String.Hashtbl.t =
 
 (* Set hook *)
 let () =
+  let default_load = !Persistent_env.Persistent_signature.load in
   Persistent_env.Persistent_signature.load :=
     fun ~unit_name ->
-      String.Hashtbl.find preloaded_cmis (String.capitalize_ascii unit_name)
+      match
+        String.Hashtbl.find preloaded_cmis (String.capitalize_ascii unit_name)
+      with
+      | Some v ->
+          Some v
+      | None ->
+          default_load ~unit_name
 
 let load_cmi_from_file file =
   String.Hashtbl.add
