--- patched_contracts/expruZKjVy3JbWcJmjtnvMGvRamkDhMgM3urGGdne3pkN9VKgK7VnD.original.tz
+++ patched_contracts/expruZKjVy3JbWcJmjtnvMGvRamkDhMgM3urGGdne3pkN9VKgK7VnD.patched.tz
@@ -293,7 +293,8 @@
                      { IF_LEFT
                          { RENAME @amt_slash_33 ;
                            { DIP { { DIP { DUP @storage } ; SWAP } } ; SWAP } ;
-                           PUSH @dest (contract :UnitContract unit) "tz1LBEKXaxQbd5Gtzbc1ATCwc3pppu81aWGc" ;
+                           PUSH key_hash "tz1LBEKXaxQbd5Gtzbc1ATCwc3pppu81aWGc";
+                           IMPLICIT_ACCOUNT @dest;
                            { DIP { { DIP { DUP @amt } ; SWAP } } ; SWAP } ;
                            UNIT ;
                            TRANSFER_TOKENS @op ;
@@ -396,7 +397,8 @@
                                               SWAP } } ;
                                       SWAP } ;
                                     { CDR ; CDR ; CDR ; CDR ; CDR ; CAR %inBaker } ;
-                                    SUB ;
+                                    SUB_MUTEZ ;
+                                    IF_NONE { PUSH string "negative mutez!"; FAILWITH } {} ;
                                     PAIR %inBaker ;
                                     SWAP ;
                                     PAIR %totalSupply ;
