--- patched_contracts/exprubv5oQmAUP8BwktmDgMWqTizYDJVhzHhJESGZhJ2GkHESZ1VWg.original.tz
+++ patched_contracts/exprubv5oQmAUP8BwktmDgMWqTizYDJVhzHhJESGZhJ2GkHESZ1VWg.patched.tz
@@ -245,7 +245,7 @@
                           CAR ;
                           CDR ;
                           CONTRACT %transfer (pair address (pair address nat)) ;
-                          IF_NONE { FAILWITH } {} ;
+                          IF_NONE { CDR; FAILWITH } {} ;
                           PUSH mutez 0 ;
                           DIG 3 ;
                           SELF ;
@@ -434,7 +434,7 @@
                           CAR ;
                           CDR ;
                           CONTRACT %transfer (pair address (pair address nat)) ;
-                          IF_NONE { FAILWITH } {} ;
+                          IF_NONE { CDR; FAILWITH } {} ;
                           PUSH mutez 0 ;
                           DIG 3 ;
                           SELF ;
@@ -746,7 +746,7 @@
                      CDR ;
                      CAR ;
                      CONTRACT unit ;
-                     IF_NONE { FAILWITH } {} ;
+                     IF_NONE { CDR; FAILWITH } {} ;
                      DIG 3 ;
                      UNIT ;
                      TRANSFER_TOKENS ;
@@ -757,7 +757,7 @@
                            CAR ;
                            CDR ;
                            CONTRACT %transfer (pair address (pair address nat)) ;
-                           IF_NONE { FAILWITH } {} ;
+                           IF_NONE { CDR; FAILWITH } {} ;
                            PUSH mutez 0 ;
                            DIG 3 ;
                            DIG 4 ;
@@ -888,7 +888,7 @@
                          CAR ;
                          CDR ;
                          CONTRACT %transfer (pair address (pair address nat)) ;
-                         IF_NONE { FAILWITH } {} ;
+                         IF_NONE { CDR; FAILWITH } {} ;
                          PUSH mutez 0 ;
                          DIG 3 ;
                          DIG 5 ;
@@ -1028,7 +1028,7 @@
                                CAR ;
                                CDR ;
                                CONTRACT %transfer (pair address (pair address nat)) ;
-                               IF_NONE { FAILWITH } {} ;
+                               IF_NONE { CDR; FAILWITH } {} ;
                                PUSH mutez 0 ;
                                DIG 3 ;
                                SELF ;
@@ -1168,7 +1168,7 @@
                      CAR ;
                      CAR ;
                      CONTRACT %xtzToToken (pair address (pair nat timestamp)) ;
-                     IF_NONE { FAILWITH } {} ;
+                     IF_NONE { CDR; FAILWITH } {} ;
                      DIG 2 ;
                      DIP 3 { DUP } ;
                      DIG 3 ;
@@ -1188,7 +1188,7 @@
                      CAR ;
                      CDR ;
                      CONTRACT %transfer (pair address (pair address nat)) ;
-                     IF_NONE { FAILWITH } {} ;
+                     IF_NONE { CDR; FAILWITH } {} ;
                      PUSH mutez 0 ;
                      DIG 3 ;
                      SELF ;
@@ -1241,7 +1241,7 @@
                          CAR ;
                          CDR ;
                          CONTRACT %getBalance (pair address (contract nat)) ;
-                         IF_NONE { FAILWITH } {} ;
+                         IF_NONE { CDR; FAILWITH } {} ;
                          PUSH mutez 0 ;
                          SELF %updateTokenPoolInternal ;
                          SELF ;
