diff --git a/src/lib_protocol_environment/environment_V3.ml b/src/lib_protocol_environment/environment_V3.ml
index c7d73dca78..b4af18ab90 100644
--- a/src/lib_protocol_environment/environment_V3.ml
+++ b/src/lib_protocol_environment/environment_V3.ml
@@ -756,4 +756,9 @@ struct
   module RPC_arg = Tezos_rpc.Arg
   module RPC_path = Tezos_rpc.Path
   module RPC_query = Tezos_rpc.Query
+
+  module Hack = struct
+    let printf = Format.printf
+  end
+
   module RPC_service = Tezos_rpc.Service

   module RPC_answer = struct
diff --git a/src/lib_protocol_environment/sigs/v3.in.ml b/src/lib_protocol_environment/sigs/v3.in.ml
index 9122fe44aa..b38ae9c127 100644
--- a/src/lib_protocol_environment/sigs/v3.in.ml
+++ b/src/lib_protocol_environment/sigs/v3.in.ml
@@ -42,6 +42,8 @@ module type T = sig
 
   module FallbackArray : [%sig "v3/fallbackArray.mli"]
 
+  module Hack : [%sig "v3/hack.mli"]
+
   module Error_monad : [%sig "v3/error_monad.mli"]
 
   open Error_monad
diff --git a/src/lib_protocol_environment/sigs/v3/hack.mli b/src/lib_protocol_environment/sigs/v3/hack.mli
new file mode 100644
index 000000000..625eeafbf
--- /dev/null
+++ b/src/lib_protocol_environment/sigs/v3/hack.mli
@@ -0,0 +1,1 @@
+val printf : ('a, Format.formatter, unit) format -> 'a
--
