--- patched_contracts/exprteUgjaxjf3jTheaj2fQkQsPFynyj3pCJ4mbzP26D4giEjQBwFb.original.tz
+++ patched_contracts/exprteUgjaxjf3jTheaj2fQkQsPFynyj3pCJ4mbzP26D4giEjQBwFb.patched.tz
@@ -1,4 +1,4 @@
-{ parameter %main unit ;
+{ parameter (unit %main) ;
   storage
     (pair :storage
        (big_map :depositors address mutez)
@@ -241,7 +241,8 @@
                    DIP 8 { DUP @s } ;
                    DIG 8 ;
                    { CDR ; CDR ; CDR ; CDR ; CDR ; CAR %deposited } ;
-                   SUB ;
+                   SUB_MUTEZ ;
+                   IF_NONE {UNIT ; FAILWITH} {};
                    PAIR %deposited ;
                    SWAP ;
                    PAIR %collat_coeff ;
@@ -344,7 +345,8 @@
                    DIG 4 ;
                    DIP 9 { DUP @user_balance } ;
                    DIG 9 ;
-                   SUB ;
+                   SUB_MUTEZ ;
+                   IF_NONE {UNIT ; FAILWITH} {};
                    ADD @withdraw_amount ;
                    UNIT ;
                    TRANSFER_TOKENS @op_withdraw ;
@@ -444,7 +446,8 @@
                         { CDR ; CDR ; CDR ; CDR ; CDR ; CDR ; CAR %borrowed } ;
                         DIP 2 { DUP @max_borrowing } ;
                         DIG 2 ;
-                        SUB ;
+                        SUB_MUTEZ ;
+                        IF_NONE {UNIT ; FAILWITH} {} ;
                         PAIR ;
                         DUP ;
                         CAR @a ;
@@ -610,7 +613,8 @@
                                   DIP 8 { DUP @s } ;
                                   DIG 8 ;
                                   { CDR ; CDR ; CDR ; CDR ; CDR ; CDR ; CAR %borrowed } ;
-                                  SUB ;
+                                  SUB_MUTEZ ;
+                                  IF_NONE {UNIT ; FAILWITH} {} ;
                                   PAIR %borrowed ;
                                   SWAP ;
                                   PAIR %deposited ;
