--- patched_contracts/exprtgpMFzTtyg1STJqANLQsjsMXmkf8UuJTuczQh8GPtqfw18x6Lc.original.tz
+++ patched_contracts/exprtgpMFzTtyg1STJqANLQsjsMXmkf8UuJTuczQh8GPtqfw18x6Lc.patched.tz
@@ -1,10 +1,5 @@
 { parameter (or (lambda %do unit (list operation)) (unit %default)) ;
-  storage
-    (pair key_hash
-          (contract
-             (or (option address)
-                 (or (pair (option address) (option mutez))
-                     (or mutez (or (pair (option address) (option mutez)) address)))))) ;
+  storage (pair key_hash address) ;
   code { DUP ;
          CAR ;
          IF_LEFT
@@ -28,6 +23,8 @@
                NIL operation ;
                { DIP { DIP { DUP } ; SWAP } ; SWAP } ;
                { DIP { DIP { DIP { DROP } } } } ;
+               CONTRACT (or (option address) (or (pair (option address) (option mutez)) (or mutez (or (pair (option address) (option mutez)) address))));
+               IF_SOME {} {PUSH string "Bad contract in storage"; FAILWITH};
                AMOUNT ;
                SENDER ;
                SOME ;
