--- patched_contracts/expru1ukk6ZqdA32rFYFG7j1eGjfsatbdUZWz8Mi1kXWZYRZm4FZVe.original.tz
+++ patched_contracts/expru1ukk6ZqdA32rFYFG7j1eGjfsatbdUZWz8Mi1kXWZYRZm4FZVe.patched.tz
@@ -243,7 +243,7 @@
                           CAR ;
                           CDR ;
                           CONTRACT %transfer (pair address (pair address nat)) ;
-                          IF_NONE { FAILWITH } {} ;
+                          IF_NONE { CDR; FAILWITH } {} ;
                           PUSH mutez 0 ;
                           DIG 3 ;
                           SELF ;
@@ -432,7 +432,7 @@
                           CAR ;
                           CDR ;
                           CONTRACT %transfer (pair address (pair address nat)) ;
-                          IF_NONE { FAILWITH } {} ;
+                          IF_NONE { CDR; FAILWITH } {} ;
                           PUSH mutez 0 ;
                           DIG 3 ;
                           SELF ;
@@ -720,7 +720,8 @@
                      DIP 3 { DUP } ;
                      DIG 3 ;
                      SWAP ;
-                     SUB ;
+                     SUB_MUTEZ ;
+                     IF_NONE { PUSH string "negative mutez!"; FAILWITH } {} ;
                      DIP { DUP ; CDR } ;
                      DIP { DUP ; DIP { CAR } ; CDR } ;
                      DIP { DUP ; DIP { CAR } ; CDR } ;
@@ -744,7 +745,7 @@
                      CDR ;
                      CAR ;
                      CONTRACT unit ;
-                     IF_NONE { FAILWITH } {} ;
+                     IF_NONE { CDR; FAILWITH } {} ;
                      DIG 3 ;
                      UNIT ;
                      TRANSFER_TOKENS ;
@@ -755,7 +756,7 @@
                            CAR ;
                            CDR ;
                            CONTRACT %transfer (pair address (pair address nat)) ;
-                           IF_NONE { FAILWITH } {} ;
+                           IF_NONE { CDR; FAILWITH } {} ;
                            PUSH mutez 0 ;
                            DIG 3 ;
                            DIG 4 ;
@@ -909,7 +910,7 @@
                          CAR ;
                          CDR ;
                          CONTRACT %transfer (pair address (pair address nat)) ;
-                         IF_NONE { FAILWITH } {} ;
+                         IF_NONE { CDR; FAILWITH } {} ;
                          PUSH mutez 0 ;
                          DIG 3 ;
                          DIG 5 ;
@@ -1040,7 +1041,8 @@
                          DUP ;
                          DIP { DIP { DUP ; CDR ; CDR ; CDR ; CDR } ;
                                SWAP ;
-                               SUB ;
+                               SUB_MUTEZ ;
+                               IF_NONE { PUSH string "negative mutez!"; FAILWITH } {} ;
                                DIP { DUP ; CDR } ;
                                DIP { DUP ; DIP { CAR } ; CDR } ;
                                DIP { DUP ; DIP { CAR } ; CDR } ;
@@ -1074,7 +1076,7 @@
                                CAR ;
                                CDR ;
                                CONTRACT %transfer (pair address (pair address nat)) ;
-                               IF_NONE { FAILWITH } {} ;
+                               IF_NONE { CDR; FAILWITH } {} ;
                                PUSH mutez 0 ;
                                DIG 3 ;
                                SELF ;
@@ -1217,7 +1219,8 @@
                      DIP 2 { DUP } ;
                      DIG 2 ;
                      SWAP ;
-                     SUB ;
+                     SUB_MUTEZ ;
+                     IF_NONE { PUSH string "negative mutez!"; FAILWITH } {} ;
                      DIP { DUP ; CDR } ;
                      DIP { DUP ; DIP { CAR } ; CDR } ;
                      DIP { DUP ; DIP { CAR } ; CDR } ;
@@ -1240,7 +1243,7 @@
                      CAR ;
                      CAR ;
                      CONTRACT %xtzToToken (pair address (pair nat timestamp)) ;
-                     IF_NONE { FAILWITH } {} ;
+                     IF_NONE { CDR; FAILWITH } {} ;
                      DIG 2 ;
                      DIP 3 { DUP } ;
                      DIG 3 ;
@@ -1260,7 +1263,7 @@
                      CAR ;
                      CDR ;
                      CONTRACT %transfer (pair address (pair address nat)) ;
-                     IF_NONE { FAILWITH } {} ;
+                     IF_NONE { CDR; FAILWITH } {} ;
                      PUSH mutez 0 ;
                      DIG 3 ;
                      SELF ;
@@ -1313,7 +1316,7 @@
                          CAR ;
                          CDR ;
                          CONTRACT %getBalance (pair address (contract nat)) ;
-                         IF_NONE { FAILWITH } {} ;
+                         IF_NONE { CDR; FAILWITH } {} ;
                          PUSH mutez 0 ;
                          SELF %updateTokenPoolInternal ;
                          SELF ;
