commit cdf9c9c25cab3824806ab10d1759f9b017705984
Author: Nic Volanschi <nic.volanschi@nomadic-labs.com>
Date:   Wed May 10 12:14:15 2023 +0200

    REMOVEME: doc patchs for old protocols

diff --git a/src/proto_017_PtNairob/lib_sc_rollup_node/dal_pages_request.mli b/src/proto_017_PtNairob/lib_sc_rollup_node/dal_pages_request.mli
index 3be2cdb1f5..f27d737fbc 100644
--- a/src/proto_017_PtNairob/lib_sc_rollup_node/dal_pages_request.mli
+++ b/src/proto_017_PtNairob/lib_sc_rollup_node/dal_pages_request.mli
@@ -42,7 +42,7 @@ type error += Dal_slot_not_found_in_store of Dal.Slot.Header.id
     The function returns [Dal_slot_not_found_in_store] if no entry is found in
     the store for the given ID (i.e. no page is registered with or without content).
 
-    If the returned value is [Some pages]], the slot whose ID is given is
+    If the returned value is [Some pages], the slot whose ID is given is
     supposed to be confirmed and [pages] correspond to the pages of the slot.
     Otherwise [None] is returned.
 
diff --git a/src/proto_017_PtNairob/lib_sc_rollup_node/inbox.mli b/src/proto_017_PtNairob/lib_sc_rollup_node/inbox.mli
index a723ad47e5..29caafad18 100644
--- a/src/proto_017_PtNairob/lib_sc_rollup_node/inbox.mli
+++ b/src/proto_017_PtNairob/lib_sc_rollup_node/inbox.mli
@@ -56,7 +56,7 @@ val start : unit -> unit Lwt.t
 
 (** [add_messages ~is_migration_block ~predecessor_timestamp
     ~predecessor inbox messages] adds [messages] to the [inbox] using
-    {!Inbox.add_all_messages}. *)
+    {!Sc_rollup.Inbox.add_all_messages}. *)
 val add_messages :
   is_migration_block:bool ->
   predecessor_timestamp:Timestamp.time ->
